rts/motoko-rts/src/persistence.rs

```diff
@@ -269,3 +281,53 @@ pub unsafe extern "C" fn buffer_in_32_bit_range() -> usize {
     assert!(size_of::<PersistentMetadata>().to_bytes().as_usize() + BUFFER_SIZE < METADATA_RESERVE);
     METADATA_ADDRESS + METADATA_RESERVE - BUFFER_SIZE
 }
+
+/// Accessor method for the weak reference registry.
+pub(crate) unsafe fn get_weak_ref_registry<M: Memory>(mem: &mut M) -> &'static mut MarkStack {
+    debug_assert!((*PersistentMetadata::get()).weak_ref_registry.get_raw() != 0);
+
+    // Lazy initialization of the weak reference registry.
+    if is_weak_ref_registry_null() {
+        // This can be run during an increment from a version of the RTS without weak reference
+        // support. In this case, the weak reference registry is NULL_POINTER (see the initial migration
+        // in function `initialize_memory`).
+        // We need to properly initialize it so that it can be used if needed.
+        // Also certain assertions in the weak reference code rely on an existing
+        // weak reference registry on which we can call is_empty().
+        initialize_weak_ref_registry(mem);
+    }
+
+    let metadata = PersistentMetadata::get();
+    let registry_value = (*metadata).weak_ref_registry;
+    let markstack_ptr = registry_value.get_ptr() as *mut MarkStack;
+    &mut *markstack_ptr
+}
+
+/// Initialize the weak reference registry in persistent metadata.
+unsafe fn initialize_weak_ref_registry<M: Memory>(mem: &mut M) {
+    debug_assert!((*PersistentMetadata::get()).weak_ref_registry == NULL_POINTER);
+    // Allocate a pointer to a MarkStack object explicitly on the heap, through a blob.
+    // No barrier needed, the lifetime of the whole weak reference registry
+    // is just during the GC marking phase. After this it can be garbage collected.
+    // Every GC phase will reinitialize the weak reference registry.
+    let markstack_blob = alloc_blob(mem, TAG_BLOB_B, Bytes(size_of::<MarkStack>()));
+    let markstack_ptr = markstack_blob.as_blob_mut() as *mut MarkStack;
+    // Initialize the MarkStack object inside this pointer.
+    *markstack_ptr = MarkStack::new(mem);
+
+    let metadata = PersistentMetadata::get();
+    (*metadata).weak_ref_registry = Value::from_ptr(markstack_ptr as usize);
```

Comment:
Maybe just a quick comment that write barrier is not needed because it is transitional (no marking needed) and not moved (no forwarding needed).